### ISSUE TITLE
Just wannting to try a simple thing builling a regression network copied from test samples

### ISSUE THREAD

**Issue by IshanSeidenberg** [mlpack/mlpack#1480]
Instead of a logistical regression i want to to simple regression so i to took out the last LogSoftMax layer output. of the (vanilla net with logistic activation function) Is this logical ? i have 3 outputs and 9 inputs an a million samples(parameter of fluid flow)! even if i have segfault somewhere i would like to be assured about this logic. thankx

-  FFN<NegativeLogLikelihood<>,RandomInitialization > model; 

-  model.Add<Linear<> >(trainData.n_rows, layer1_size);

-   model.Add<SigmoidLayer<> >();

-  model.Add<Linear<> >(layer1_size, outputSize);

-  //model.Add<LogSoftMax<> >(); // left out

- RMSProp opt(0.01, 32, 0.88, 1e-8, maxEpochs * trainData.n_cols, -1);

**Comment by zoq** [comment]
The negative log-likelihood loss function expects that the input contains log-probabilities, which can be obtained by adding a ``LogSoftMax`` layer, or you could switch to the ``CrossEntropyError`` function if you don't like the idea to add another layer. However, since you like to do regression a common choice is to use the ``MeanSquaredError`` instead of NegativeLogLikelihood. You can test your model with a linear function as the last layer, using a ReLu should work as well. I hope this is helpful, let me know if I should clarify anything.

> even if i have segfault somewhere i would like to be assured about this logic

Do you think you could provide a simple example to reproduce the issue?

**Comment by IshanSeidenberg** [comment]
i understood the MeanSquaredError.
i don't have the seffault anymore!
here is the code:
i tested x2 
training mode: 1000 points form 0 to 2
test mode : 1000 point from 0. to 1.5
i dont think it worked.

`#include <iostream>
#include <fstream>
#include <cmath>
#include <mlpack/core.hpp>

#include <mlpack/methods/ann/layer/layer.hpp>
#include <mlpack/methods/ann/layer/layer_types.hpp>
#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
#include <mlpack/methods/ann/loss_functions/sigmoid_cross_entropy_error.hpp>
#include <mlpack/methods/ann/init_rules/random_init.hpp>
#include <mlpack/methods/ann/init_rules/const_init.hpp>
#include <mlpack/methods/ann/init_rules/nguyen_widrow_init.hpp>
#include <mlpack/methods/ann/ffn.hpp>
#include <mlpack/methods/ann/rnn.hpp>

using namespace mlpack;
using namespace std;
using namespace ann;
using namespace mlpack::ann;
using namespace optimization;
void pinfo(arma::mat m, const char *txt)
{
        printf("%s(%d,%d)\n",txt,(int)m.n_rows,(int)m.n_cols);
}
void savemat(arma::mat m, const char*filename)
{

	FILE *ft;
	ft = fopen(filename,"w");
	for(int i=0;i<m.n_rows;i++)
	{
		for(int j=0;j<m.n_cols;j++)
    		fprintf(ft,"%e ",m(i,j));//i,i);
		fprintf(ft,"\n");//i,i);
	}
	
fclose(ft);
}
void saveerr(arma::mat measured,arma::mat expected,const char*filename)
{
    double err = 0.0;
    double diff_abs = 0.0;
    double mean_measured = 0.0;
    double mean_expected = 0.0;
    double diff_measured = 0.0;
    double diff_expected = 0.0;
    double diff_err = 0.0;
    double mean2 = 0.0;
    FILE *f;
    f = fopen(filename,"w");

    for(int e=0;e<m.n_elem;e++)
    {
        mean_measured  += measured(e);
        mean_expected  += expected(e);
    }

    mean_measured /= measured.n_elem;
    mean_expected /= expected.n_elem;
    mean2 = mean_measured*mean_expected;
    for(int e=0;e<measured.n_elem;e++)
    {
        diff_abs += abs(measured(e) - expected(e));
        diff_measured = abs(measured(e) - mean_expected);
        diff_expected = abs(expected(e) - mean_measured);
        diff_err += (diff_m*diff_expected)/mean2;
        fprintf(f,"%e <- %e : absolute difference\n", measured(e), expteded(e), diff_abs);
    }

    fprintf(f,"mean_abs_error : %e\n", diff_abs/measured.n_elem);
    fprintf(f,"mean_mesured = %e\n",mean_measured);
    fprintf(f,"mean_expected = %e\n",mean_expected);
    fprintf(f,"diff_m*diff_e/mean_a*mean_b",diff_expected*diff_measured/mean_expected*mean_measured);

    printf("mean_error : %e\n", err/measured.n_elem);
    printf("mean_m = %e\n",mean_measured);
    printf("mean_a = %e\n",mean_expected);
    printf("diff_a*diff_b/mean_a*mean_b",diff_a*diff_m/mean_expected*mean_measured);

    fclose(f);
}

void load_dataset(const char* filename,int cut,arma::mat *trainData_,arma::mat *trainLabels_)
{
  arma::mat dataset;
  //arma::mat trainData = *trainData_;
  //arma::mat trainLabels = *trainLabels_;
  
  arma::mat trainData; //= *trainData_;
  arma::mat trainLabels;// = *trainLabels_;
  
  data::Load(filename, dataset, true);

  trainData = dataset.submat(0, 0, dataset.n_rows - (cut +1),
  dataset.n_cols - 1);

  arma::mat trainLabelsTemp = dataset.submat(dataset.n_rows - cut , 0,
      dataset.n_rows - 1, dataset.n_cols - 1);
      
  trainLabels = arma::zeros<arma::mat>(1, trainLabelsTemp.n_cols);
  for (size_t i = 0; i < trainLabelsTemp.n_cols; ++i)
  {
   (trainLabels(i)) = arma::as_scalar(arma::find(
        arma::max(trainLabelsTemp.col(i)) == trainLabelsTemp.col(i), 1)) + 1;
  }
  
  *trainData_ = trainData;
  *trainLabels_ =  trainLabels;
  
}


/**
 * Train and evaluate a vanilla network with the specified structure.
 */
template<typename MatType = arma::mat>
void BuildVanillaNetwork(MatType& trainData,
                         MatType& trainLabels,
                         MatType& testData,
                         MatType& testLabels,                  
                         const size_t hiddenLayerSize,
                         const size_t maxEpochs,
                         const double classificationErrorThreshold)
{
  /*
   * Construct a feed forward network with trainData.n_rows input nodes,
   * hiddenLayerSize hidden nodes and trainLabels.n_rows output nodes.
   */

    FFN<MeanSquaredError<>> model;
    model.Add<Linear<> >(trainData.n_rows,hiddenLayerSize);// hiddenLayerSize);
    model.Add<SigmoidLayer<>>();
    model.Add<Linear<> >(hiddenLayerSize, trainLabels.n_rows);
    model.Add<ReLULayer<>>();

     // optimizer
    double  	stepSize = 0.001;
    size_t  	batchSize = 1;
    const double  	beta1 = 0.9;
    const double  	beta2 = 0.999;
    const double  	eps = 1e-8;
    size_t  	maxIterations = trainData.n_cols;
    double  	tolerance = 1e-5;
    bool  	shuffle = true;

    optimization::StandardSGD opt1(stepSize,batchSize, maxEpochs *maxIterations,tolerance,shuffle);
    optimization::RMSProp  opt2(0.01, 1, 0.88, 1e-8, maxEpochs * trainData.n_cols, -1);
    printf("---model.Train()-----\n");
    pinfo(trainData,"trainData");
    pinfo(trainLabels,"trainLabels");
    model.Train(trainData, trainLabels,opt1);
    printf("---model.Prediction()-----\n");
    MatType trainPrediction;
    model.Predict(trainData, trainPrediction);
    printf("model.autotest\n");
    saveerr(trainPrediction,trainLabels,"error.train.dat");

    MatType testPrediction;
    model.Predict(testData, testPrediction);
    printf("---model.test()-----\n");
    saveerr(testPrediction,testLabels,"error.test.dat");
}

/**
 * Train the vanilla network on x^2.
 */
int main(int argc,char**argv)
{

  arma::mat trainData;
  data::Load("x2.train.data.csv", trainData, true); 
  pinfo(trainData,"trainData");
  
  arma::mat trainLabels;
  data::Load("x2.train.label.csv", trainLabels, true);
  pinfo(trainLabels,"trainLabels");
  
  arma::mat testData;
  data::Load("x2.test.data.csv", testData, true); 
  pinfo(testData,"testData");
  
  arma::mat testLabels;
  data::Load("x2.test.label.csv", testLabels, true);
  pinfo(testLabels,"testLabels");

  // Vanilla neural net with ReLULayer
  BuildVanillaNetwork<>
      (trainData, trainLabels, testData, testLabels, 32, 20, 0.01);
}
`

**Comment by zoq** [comment]
``LogSoftmax`` applies logarithm after softmax, softmax squashes a vector of a specific size 0 and 1, so that we could interpret the output of the softmax as the probabilities. But that also means we can't put the softmax layer at the beginning, it should be placed at the end of the network. ``CrossEntropyError`` does the transformation inside the loss function, so you don't have to add a LogSoftmax layer (isntead of using ``NegativeLogLikelihood + LogSoftmax`` you can use ``CrossEntropyError``). However, since you like to do regression, probabilities isn't what you like to see at the end, so using ``MeanSquaredError`` is probably the best option.

```c++
FFN<MeanSquaredError<>> model;
model.Add<Linear<> >(trainData.n_rows,hiddenLayerSize);// hiddenLayerSize);
model.Add<SigmoidLayer<>>();
model.Add<Linear<> >(hiddenLayerSize, trainLabels.n_rows);
model.Add<ReLULayer<>>();
```


just looked at the architecture for now, which looks reasonable; you could test if adding more layers does have an effect, also perhaps you could use   ReLU function for the second layer as well, it might help to reduce the likelihood of the gradient to vanish.

**Comment by rcurtin** [comment]
I've closed this issue for inactivity.  Please feel free to reopen it if you are still having problems! :)
